core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java

```diff
@@ -467,12 +571,54 @@ public final void close(Throwable cause) {
      */
     protected final boolean tryClose(Throwable cause) {
         if (setState(State.OPEN, State.CLOSED)) {
-            addObjectOrEvent(new CloseEvent(cause));
+            if (!tryCollect(cause)) {
+                addObjectOrEvent(new CloseEvent(cause));
+            }
             return true;
         }
         return false;
     }
 
+    private boolean tryCollect(@Nullable Throwable cause) {
+        final SubscriptionImpl subscription = this.subscription;
+        if (subscription != null) {
+            @SuppressWarnings("unchecked")
+            final CompletableFuture<List<T>> collectingFuture =
+                    (CompletableFuture<List<T>>) subscription.collectingFuture();
+            if (collectingFuture != null) {
+                if (setState(State.CLOSED, State.CLEANUP)) {
+                    if (subscription.needsDirectInvocation()) {
+                        tryCollect0(cause, subscription, collectingFuture);
+                    } else {
+                        subscription.executor().execute(() -> {
+                            tryCollect0(cause, subscription, collectingFuture);
+                        });
+                    }
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private void tryCollect0(@Nullable Throwable cause, SubscriptionImpl subscription,
```

Comment:
nit: `0` is not necessary because it has different signature from `tryCollect()` above.